Bring i18n functionality up to date

swissspidy opened the pull request:
## Description

This PR does the following:

* Removes unnecessary `load_plugin_textdomain()  call  
  This is only needed for plugins that need to run on WordPress < 4.6 or ship their own translation files. Gutenberg does neither.
* Removes calls to `pot-to-php` when generating the ZIP file, instead relying on WordPress.org to parse JavaScript files (see https://meta.trac.wordpress.org/ticket/3748)
* Uses `wp_set_script_translations` when overriding scripts to rely on JSON translation files provided by WordPress.org

Tagged `[WIP]` because the JSON translation files don't seem to be generated reliably on WordPress.org yet because of some memory spikes. See https://wordpress.slack.com/archives/C02QB8GMM/p1542805941361400 

Fixes #12549.

## How has this been tested?

Tested by building the ZIP file, installing the plugin with it and downloading current translations from WordPress.org.

Also tested by running `wp i18n make-pot` against the extracted ZIP file.


## Types of changes
Bug fixes under the hood.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

aduth commented on languages/README.md:
```diff
@@ -1,10 +0,0 @@
-Languages
```
We also need to remove?

https://github.com/WordPress/gutenberg/blob/2fc34bc151bd53b72dc9a5fb7d4112be3ae8828c/babel.config.js#L19-L24

swissspidy replied:
Yes, good point! I didn‘t notice it was there.

aduth replied:
And these? From a search of "languages/"

https://github.com/WordPress/gutenberg/blob/2fc34bc151bd53b72dc9a5fb7d4112be3ae8828c/.gitignore#L7-L8

https://github.com/WordPress/gutenberg/blob/2fc34bc151bd53b72dc9a5fb7d4112be3ae8828c/bin/build-plugin-zip.sh#L121-L122

swissspidy replied:
Yep, to be removed as well.

swissspidy commented on lib/rest-api.php:
```diff
@@ -317,36 +317,3 @@ function gutenberg_silence_rest_errors() {
- * @param string $post_type Post type key.
- * @return array Arguments supplied to register_post_type()
- */
-function gutenberg_filter_post_type_labels( $args, $post_type ) {
```
Related: #10656.

aduth commented:
I see a "memory exhausted" warning on this branch, which may explain the current build failures.

Note: I force pushed the branch as well, so you may need to:

```
git checkout master
git fetch origin
git branch -D i18n-updates
git checkout i18n-updates
```

aduth commented on lib/client-assets.php:
```diff
@@ -111,6 +111,10 @@ function register_tinymce_scripts() {
 	wp_register_script( $handle, $src, $deps, $ver, $in_footer );
+
+	if ( function_exists( 'wp_set_script_translations' ) ) {
+		wp_set_script_translations( $handle, 'gutenberg' );
```
Commenting this line resolves the memory exhaustion for me. Perhaps getting stuck in some infinite recursion where `wp_set_script_translations` is in-turn causing `gutenberg_override_script` to be called once more?

aduth commented on lib/client-assets.php:
```diff
@@ -111,6 +111,10 @@ function register_tinymce_scripts() {
 	wp_deregister_script( $handle );
 	wp_register_script( $handle, $src, $deps, $ver, $in_footer );
+
+	if ( function_exists( 'wp_set_script_translations' ) ) {
```
Do we need `function_exists` at this point? Assuming we can change the minimum required version for the plugin to 5.0+, which seems a reasonable expectation.

aduth commented on lib/i18n.php:
```diff
@@ -38,17 +38,3 @@ function gutenberg_get_jed_locale_data( $domain ) {
 
```
I might expect we should be able to do away with this file altogether? Do we need `gutenberg_get_jed_locale_data` ? It's still used in `client-assets.php`, but I would assume we should be able to rely on the core behavior now to bootstrap the localized data.

aduth commented:
Memory exhaustion occurs here:

https://github.com/WordPress/WordPress/blob/4e3e9ca9cf893190609728848caf224034fd1ef0/wp-includes/class.wp-dependencies.php#L338

I think it stems from this line:

https://github.com/WordPress/WordPress/blob/4e3e9ca9cf893190609728848caf224034fd1ef0/wp-includes/class.wp-scripts.php#L517

Where `wp-i18n` is adding itself as a dependency when we're overriding the core copy.

swissspidy commented:
Thanks for the comments, I‘ll have another look at it. Didn‘t realize the issues might be because of recursion.

aduth committed:
Bring i18n functionality up to date

aduth commented:
I rebased and force-pushed a new set of updates, including deprecations / removals of what I think we should be able to be rid of with the custom `gutenberg_load_locale_data` behavior. In testing though, it doesn't appear to be working. I'm having a hard time following when / how the plugin JSON file gets downloaded, as it's not present for me in my languages directory.

Thoughts / advice? On this or otherwise on the latest changes?

aduth commented:
Oh, I also created a Trac ticket for the memory exhaustion behavior when calling `wp_set_script_translations` with `'wp-i18n'` (and, as it turns out, `'wp-polyfill'`):

https://core.trac.wordpress.org/ticket/46089

swissspidy commented:
> In testing though, it doesn't appear to be working. I'm having a hard time following when / how the plugin JSON file gets downloaded, as it's not present for me in my languages directory.

If you go to Dashboard -> Updates to download the latest translations, WordPress should get all the JSON files from WordPress.org as part of the language packs. cc @ocean90 for verification.

aduth committed:
Plugin: Add Text Domain to plugin headers

aduth commented:
Did another rebase to resolve newer conflicts. Pushed another change to add the missing `Text Domain` to the plugin headers, which is documented as a requirement for the plugin repository to pick up on the strings:

https://make.wordpress.org/core/2018/11/09/new-javascript-i18n-support-in-wordpress/

Some related discussion from Slack ([link requires registration](https://make.wordpress.org/chat/)):

https://wordpress.slack.com/archives/C5UNMSU4R/p1548882554475700

A couple highlights:

- Dropping `gutenberg-translations.php` may risk losing translator comments, since Gutenberg only distributes minified code in the plugin repository. The current build process generates a dummy PHP file from the source files.
- It's not clear that the JSON strings are being produced or otherwise made available from the plugin repository (for any plugin). If true, it may be sensible to consider shipping these with the plugin, depending on whether there's any resolution to the prior point.

aduth commented:
Current Status: After subsequent conversation from the [above-mentioned Slack link](https://wordpress.slack.com/archives/C5UNMSU4R/p1548882554475700), it was concluded that it's best that this pull request (and generally the effort of Gutenberg to leverage `wp_set_script_translations`) be put on hold until a point where issues with the plugin repository generating locale data can be resolved, as otherwise it would regress in the plugin's ability to localize its interface.

gziolo commented:
To reflect @aduth's comment I labeled this PR as blocked to make triaging and reviewing easier.

ocean90 commented:
The [German language pack](https://downloads.wordpress.org/translation/plugin/gutenberg/5.1.1/de_DE.zip) now includes translations in JSON format.

aduth commented:
Related:

- https://meta.trac.wordpress.org/changeset/8338/
- https://meta.trac.wordpress.org/ticket/3876

aduth commented:
> The [German language pack](https://downloads.wordpress.org/translation/plugin/gutenberg/5.1.1/de_DE.zip) now includes translations in JSON format.

Is this locale specific? Or is it just a matter of needing to publish a new copy of the language pack for the changes to take effect?

For example, the [French language pack](https://downloads.wordpress.org/translation/plugin/gutenberg/5.1.1/fr_FR.zip) does _not_ include the JSON files.

swissspidy commented:
@aduth According to https://api.wordpress.org/translations/plugins/1.0/?slug=gutenberg&version=5.1.1 the French translation for Gutenberg hasn't been updated in a while, thus there is no new language pack yet. So yeah, I guess it's just a matter of pressing a button :-)

aduth commented:
In refreshing this, there's a remaining challenge in how we consider the domain for Gutenberg's strings.

As I see it, there are a few options, though the implementation specifics of each are not quite so clear:

- Set the plugin strings as the `'gutenberg'` domain, and either:
   - Transform the translation function calls in the source to reference the plugin domain in a way which doesn't impact core (e.g. Babel transform to augment the arguments, or monkey-patching `wp.i18n.__` and friends in the context of the plugin)
   -  Trick the [translation strings printing](https://github.com/WordPress/wordpress-develop/blob/3d246f985ab4846493168cb879c9b0214c4c0ba1/src/wp-includes/class.wp-scripts.php#L550-L554) to output the Gutenberg strings data as the default domain
- Set the plugin strings as the `'default'` domain, and in some way trick the [script textdomain loading](https://github.com/WordPress/wordpress-develop/blob/3d246f985ab4846493168cb879c9b0214c4c0ba1/src/wp-includes/l10n.php#L912) into loading the strings from the plugin translation files

aduth committed:
Plugin: Set Gutenberg script translations as default domain

Filter loading behavior to load from plugin translation files

aduth commented:
I've pushed a rebased branch. In addressing the [prior-noted challenge](https://github.com/WordPress/gutenberg/pull/12559#issuecomment-467934557), I chose in 299adc6 to alter the load behavior of `load_script_translations`. It's not the most elegant, but I think it mimics well the prior plugin behavior (assume default domain). And, importantly, it actually works:

![image](https://user-images.githubusercontent.com/1779930/53509592-17c20580-3a8a-11e9-8acb-10d7a6e48d8d.png)

aduth committed:
Testing: Update gutenberg_override_script per localization changes

aduth committed:
Plugin: Provide second argument for set script translations

Technically optional, only after WP5.1+, which is newer than the current minimum version supported by Gutenberg

aduth commented on babel.config.js:
```diff
@@ -3,17 +3,5 @@ module.exports = function( api ) {
-			production: {
-				plugins: [
-					[
-						'@wordpress/babel-plugin-makepot',
```
Arguably I wonder if it's still necessary to include `@wordpress/babel-plugin-makepot` as a dependency in `package.json`, since it's no longer referenced at all anywhere in the project. But I think it's probably simpler / non-harmful to just have `package.json` include _every single package_ we distribute.

aduth commented:
> * Removes calls to `pot-to-php` when generating the ZIP file, instead relying on WordPress.org to parse JavaScript files (see https://meta.trac.wordpress.org/ticket/3748)

Should we have removed all of the `pot-to-php` related files? I see they're still present in the code, but don't appear to be referenced anywhere anymore.
